core/src/test/java/org/apache/iceberg/actions/TestBinPackRewriteFileGroupPlanner.java

```diff
@@ -0,0 +1,468 @@
+    planner.init(
+        ImmutableMap.of(
+            BinPackRewriteFileGroupPlanner.REWRITE_ALL,
+            "true",
+            BinPackRewriteFileGroupPlanner.MAX_FILE_GROUP_SIZE_BYTES,
+            "10"));
+
+    FileRewritePlan<FileGroupInfo, FileScanTask, DataFile, RewriteFileGroup> plan = planner.plan();
+
+    assertThat(plan.totalGroupCount()).isEqualTo(6);
+    assertThat(plan.groupsInPartition(FILE_1.partition())).isEqualTo(3);
+    assertThat(plan.groupsInPartition(FILE_4.partition())).isEqualTo(2);
+    assertThat(plan.groupsInPartition(FILE_6.partition())).isEqualTo(1);
+  }
+
+  @Test
+  void testFilter() {
+    addFiles();
+    BinPackRewriteFileGroupPlanner planner =
+        new BinPackRewriteFileGroupPlanner(
+            table,
+            Expressions.or(
+                Expressions.equal(Expressions.bucket("data", 16), 0),
+                Expressions.equal(Expressions.bucket("data", 16), 2)));
+    planner.init(REWRITE_ALL);
+
+    FileRewritePlan<FileGroupInfo, FileScanTask, DataFile, RewriteFileGroup> plan = planner.plan();
+
+    List<RewriteFileGroup> groups = Lists.newArrayList(plan.groups().iterator());
+    assertThat(plan.totalGroupCount()).isEqualTo(2);
+    assertThat(groups).hasSize(2);
+    assertThat(groups.stream().mapToLong(RewriteFileGroup::inputFileNum).sum()).isEqualTo(4);
+  }
+
+  @Test
+  void testMaxOutputFileSize() {
+    addFiles();
+    int targetFileSize = 10;
+    BinPackRewriteFileGroupPlanner planner = new BinPackRewriteFileGroupPlanner(table);
+    planner.init(
+        ImmutableMap.of(
+            BinPackRewriteFileGroupPlanner.REWRITE_ALL,
+            "true",
+            BinPackRewriteFileGroupPlanner.TARGET_FILE_SIZE_BYTES,
+            String.valueOf(targetFileSize)));
+
+    FileRewritePlan<FileGroupInfo, FileScanTask, DataFile, RewriteFileGroup> plan = planner.plan();
+
+    assertThat(plan.groups().iterator().next().maxOutputFileSize())
+        .isGreaterThan(targetFileSize)
+        .isLessThan((long) (targetFileSize * MAX_FILE_SIZE_DEFAULT_RATIO));
+  }
+
+  @Test
+  void testExpectedOutputFiles() {
+    addFiles();
+    BinPackRewriteFileGroupPlanner planner = new BinPackRewriteFileGroupPlanner(table);
+    planner.init(
+        ImmutableMap.of(
+            BinPackRewriteFileGroupPlanner.REWRITE_ALL,
+            "true",
+            BinPackRewriteFileGroupPlanner.TARGET_FILE_SIZE_BYTES,
+            "21"));
+
+    FileRewritePlan<FileGroupInfo, FileScanTask, DataFile, RewriteFileGroup> plan = planner.plan();
+
+    List<RewriteFileGroup> groups = Lists.newArrayList(plan.groups().iterator());
+    assertThat(groups).hasSize(3);
+    for (RewriteFileGroup group : groups) {
+      if (FILE_1.partition().equals(group.info().partition())) {
+        assertThat(group.expectedOutputFiles()).isEqualTo(2);
+      } else if (FILE_4.partition().equals(group.info().partition())) {
+        assertThat(group.expectedOutputFiles()).isEqualTo(1);
```

Comment:
is this the case to avoid the 2nd small remainder file? otherwise, group size of 22 is larger than the target size of 21.

Reply:
This one is a bit tricky.
We allow file size 10% bigger than the target. See: `SizeBasedFileRewritePlanner.expectedOutputFiles` method
```
   * [..] If the new average file size is no more than 10% greater than
   * the target file size, then this method will round down when determining the number of output
   * files. Otherwise, the remainder will be written into a separate file.
```

Added some comment to describe the tests case

Reply:
yes, that is the scenario I was asking. thx for confirming and clarifying in the code